modules/apps/asset/asset-categories-selector-web/src/main/resources/META-INF/resources/js/SelectCategory.es.js

```diff
@@ -192,6 +192,13 @@ function SelectCategory({
 						{flattenedNodes.length > 0 ? (
 							<Treeview
 								NodeComponent={Treeview.Card}
+								filter={(node, filterQuery) =>
+									!node.vocabulary &&
+									node.name
+										.toLowerCase()
+										.indexOf(filterQuery.toLowerCase()) !==
+										-1
+								}
```

Comment:
Just thinking out loud here, but the API has a bit of a "redundant" feel to it, and not just because the word `filter` appears in two props (which might be confusing for people and definitely requires documentation, unless you think it will be obvious to people what `filter` and `filterQuery` are, and how to use them, and when).

The thing is, the callsite here has access to `filterQuery` already, and it's then passing in a function that will get that same `filterQuery` passed back into it, when it could have just accessed it via the closure. Among other things this function we pass in is going to call `toLowerCase()` again and again on the passed in `filterQuery` in a loop (due to the `forEach()` calls), always producing exactly the same result.

All of which is to say I wonder if we couldn't do this with a single prop (`PropTypes.oneOfType[PropTypes.func. PropTypes.string]`): if passed a string, behave exactly as before, if passed a func, use it to filter; eg.

```js
const query = filterQuery.toLowerCase();

<Treeview
    filter={(node) =>
        !node.vocabulary &&
        node.name.toLowerCase().includes(query)
    }
/>
```

Reply:
I agree this approach is better, the only thing that worries me is that is a breaking change, but I'm not sure if someone outside Liferay is using it 🤔  Should we just go for it removing the old prop? or maybe leave the prop for compatibility?

What do you think?

@jbalsas what do we usually do with breaking changes in javascript API?

Reply:
I assumed `filterQuery` was local to the treeview... based on this I guess the search input is controlled from the outside, which makes sense.

I think we could deprecate the `filterQuery` prop but keep it in this version. If `filter` is passed, we ignore `filterQuery`, otherwise it works as before.

I think it's almost certain no one is using this outside of Liferay, but we could be using it inside in different ways. The cost of keeping this for backwards compatibility seems relatively small, so I guess we can keep it around.

Reply:
> the only thing that worries me is that is a breaking change

Ah, I didn't mean to propose a breaking change — instead of replacing `filterQuery` with `filter`, I meant to augment the behavior of the existing prop in a non-breaking way; the rename is just me being careless and imprecise. (Of course, `filter` is a better name, but 🤷‍♂️ ).

> I think it's almost certain no one is using this outside of Liferay

If that's true, then maybe we shouldn't worry about this?...

> but we could be using it inside in different ways. The cost of keeping this for backwards compatibility seems relatively small, so I guess we can keep it around.

Well, that is the cost I was trying to avoid (of having two confusingly similar props) — the thing is, complexity accrues one prop at a time, and before you know it you've got a ghastly, Frankenstein API that is hard to use and hard to evolve.

Reply:
> If that's true, then maybe we shouldn't worry about this?...

I wouldn't necessarily worry about this right now... just check for internal usages...

Reply:
Is there any protocol for deprecating props like that? If none exist I think it may be better to leave the name of the property unchanged to avoid confusion 🤔  what do you think?

Reply:
Just remove it 😂

Reply:
![](https://media.giphy.com/media/3ohze1E8A3HmQD1HyM/giphy.gif)